File: server/schemas/invoke.py
"""Schemas for the /invoke endpoint — the single Headless API entry point."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class InvokeRequest(BaseModel):
    """Client sends this to talk to an agent."""

    agent_id: str
    session_id: str | None = None  # None = create new session
    user_id: str = "anonymous"
    message: str
    # Optional: pre-filled form data (for workflow steps)
    form_data: dict[str, Any] | None = None
    # Optional: explicit intent override
    intent: str | None = None
    # Whether the user wants expanded answer
    expand: bool = False
    # Client metadata (device, channel, etc.)
    client_meta: dict[str, Any] | None = None


class Citation(BaseModel):
    source_id: str
    source_name: str
    content_snippet: str
    page: int | None = None
    paragraph: int | None = None
    line_start: int | None = None
    line_end: int | None = None
    score: float | None = None


class WorkflowCard(BaseModel):
    """A UI card for a workflow step."""
    step_name: str
    step_type: str
    prompt: str
    fields: list[dict[str, Any]] | None = None
    # Current step index & total
    current_step: int = 0
    total_steps: int = 0


class InvokeResponse(BaseModel):
    """Unified response from any agent invocation."""

    session_id: str
    trace_id: str
    # Core answer
    short_answer: str
    expanded_answer: str | None = None
    # Citations with traceability
    citations: list[Citation] = Field(default_factory=list)
    # Suggested follow-up questions
    suggested_followups: list[str] = Field(default_factory=list)
    # Workflow card (if in a workflow)
    workflow_card: WorkflowCard | None = None
    # Workflow status
    workflow_status: str | None = None  # None | in_progress | waiting_input | completed | escalated
    # Escalation info
    escalated: bool = False
    escalation_reason: str | None = None
    # Metadata
    metadata: dict[str, Any] | None = None

File: console/src/types/invoke.ts
export interface InvokeRequest {
  agent_id: string;
  message: string;
  session_id?: string;
  tenant_id?: string;
  form_data?: Record<string, unknown>;
}

export interface InvokeResponse {
  session_id: string;
  trace_id: string;
  short_answer: string;
  expanded_answer?: string;
  citations: Citation[];
  suggested_followups: string[];
  workflow_card?: WorkflowCard;
  workflow_status?: string;
  escalated: boolean;
  escalation_reason?: string;
  skill_info?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface Citation {
  source_id: string;
  source_name: string;
  content_snippet: string;
  page?: number | null;
  paragraph?: number | null;
  line_start?: number | null;
  line_end?: number | null;
  score?: number | null;
}

export interface WorkflowCardField {
  name: string;
  label: string;
  field_type?: string;
  required?: boolean;
  placeholder?: string;
  options?: Array<{ label: string; value: string }>;
  file_config?: { max_size_mb?: number; allowed_extensions?: string[] };
}

export interface WorkflowCard {
  step_name: string;
  step_type: string;
  prompt: string;
  fields?: WorkflowCardField[];
  current_step: number;
  total_steps: number;
  collected_data?: Record<string, unknown>;
  webhook_result?: Record<string, unknown>;
}

export interface SSEEvent {
  event: 'status' | 'answer' | 'done' | 'error';
  data: Record<string, unknown>;
}
